File: sdk/python/src/sandboxmatrix/http_client.py
"""HTTP client for sandboxMatrix REST API."""

import json
from urllib.request import Request, urlopen
from urllib.error import HTTPError
from .models import Sandbox, ExecResult, Matrix, Session
from .exceptions import SandboxMatrixError, SandboxNotFoundError


class HTTPClient:
    """HTTP client for sandboxMatrix REST API server.

    Connects directly to the REST API server instead of wrapping the CLI binary.
    Start the server with: smx server start --addr :8080

    Args:
        base_url: Base URL of the API server (default: http://localhost:8080)
        token: Optional Bearer token for RBAC authentication
    """

    def __init__(self, base_url: str = "http://localhost:8080", token: str | None = None):
        self.base_url = base_url.rstrip("/")
        self.token = token

    def _request(self, method: str, path: str, body: dict | None = None) -> dict | list:
        url = f"{self.base_url}/api/v1{path}"
        data = json.dumps(body).encode() if body else None
        req = Request(url, data=data, method=method)
        req.add_header("Content-Type", "application/json")
        if self.token:
            req.add_header("Authorization", f"Bearer {self.token}")

        try:
            with urlopen(req) as resp:
                return json.loads(resp.read().decode())
        except HTTPError as e:
            body_text = e.read().decode() if e.fp else ""
            try:
                err_data = json.loads(body_text)
                msg = err_data.get("error", body_text)
            except json.JSONDecodeError:
                msg = body_text
            if e.code == 404:
                raise SandboxNotFoundError(msg) from e
            raise SandboxMatrixError(f"API error ({e.code}): {msg}") from e

    def health(self) -> dict:
        return self._request("GET", "/health")

    def version(self) -> dict:
        return self._request("GET", "/version")

    # Sandbox operations
    def create_sandbox(self, name: str, blueprint: str, workspace: str | None = None) -> Sandbox:
        body = {"name": name, "blueprint": blueprint}
        if workspace:
            body["workspace"] = workspace
        data = self._request("POST", "/sandboxes", body)
        return self._parse_sandbox(data)

    def get_sandbox(self, name: str) -> Sandbox:
        data = self._request("GET", f"/sandboxes/{name}")
        return self._parse_sandbox(data)

    def list_sandboxes(self) -> list[Sandbox]:
        data = self._request("GET", "/sandboxes")
        return [self._parse_sandbox(s) for s in data]

    def start_sandbox(self, name: str) -> None:
        self._request("POST", f"/sandboxes/{name}/start")

    def stop_sandbox(self, name: str) -> None:
        self._request("POST", f"/sandboxes/{name}/stop")

    def destroy_sandbox(self, name: str) -> None:
        self._request("DELETE", f"/sandboxes/{name}")

    def exec(self, name: str, command: str | list[str]) -> ExecResult:
        if isinstance(command, str):
            cmd = ["sh", "-c", command]
        else:
            cmd = command
        data = self._request("POST", f"/sandboxes/{name}/exec", {"command": cmd})
        return ExecResult(
            exit_code=data.get("exitCode", -1),
            stdout=data.get("stdout", ""),
            stderr=data.get("stderr", ""),
        )

    def stats(self, name: str) -> dict:
        return self._request("GET", f"/sandboxes/{name}/stats")

    # Snapshot operations
    def create_snapshot(self, name: str, tag: str | None = None) -> dict:
        body = {}
        if tag:
            body["tag"] = tag
        return self._request("POST", f"/sandboxes/{name}/snapshots", body)

    def list_snapshots(self, name: str) -> list:
        return self._request("GET", f"/sandboxes/{name}/snapshots")

    # Matrix operations
    def create_matrix(self, name: str, members: list[dict]) -> dict:
        return self._request("POST", "/matrices", {"name": name, "members": members})

    def get_matrix(self, name: str) -> dict:
        return self._request("GET", f"/matrices/{name}")

    def list_matrices(self) -> list:
        return self._request("GET", "/matrices")

    def start_matrix(self, name: str) -> None:
        self._request("POST", f"/matrices/{name}/start")

    def stop_matrix(self, name: str) -> None:
        self._request("POST", f"/matrices/{name}/stop")

    def destroy_matrix(self, name: str) -> None:
        self._request("DELETE", f"/matrices/{name}")

    # Session operations
    def start_session(self, sandbox: str) -> dict:
        return self._request("POST", "/sessions", {"sandbox": sandbox})

    def list_sessions(self, sandbox: str | None = None) -> list:
        path = "/sessions"
        if sandbox:
            path += f"?sandbox={sandbox}"
        return self._request("GET", path)

    def end_session(self, session_id: str) -> None:
        self._request("POST", f"/sessions/{session_id}/end")

    @staticmethod
    def _parse_sandbox(data: dict) -> Sandbox:
        metadata = data.get("metadata", {})
        spec = data.get("spec", {})
        status = data.get("status", {})
        return Sandbox(
            name=metadata.get("name", ""),
            state=status.get("state", "Unknown"),
            blueprint=spec.get("blueprintRef", ""),
            runtime_id=status.get("runtimeID", ""),
            ip=status.get("ip"),
        )

File: sdk/typescript/src/http_client.ts
import type {
  Sandbox, ExecResult, Matrix, Session, VersionInfo,
  FileInfo, PortMapping, InterpretResult, BuildResult,
} from "./models";
import { SandboxMatrixError, SandboxNotFoundError } from "./errors";

export interface HTTPClientOptions {
  baseURL?: string;
  token?: string;
}

export class HTTPClient {
  private baseURL: string;
  private token?: string;

  constructor(options: HTTPClientOptions = {}) {
    this.baseURL = (options.baseURL ?? "http://localhost:8080").replace(/\/+$/, "");
    this.token = options.token;
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const url = `${this.baseURL}/api/v1${path}`;
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.token) {
      headers["Authorization"] = `Bearer ${this.token}`;
    }

    const res = await fetch(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    });

    const data = await res.json();
    if (!res.ok) {
      const msg = data?.error ?? JSON.stringify(data);
      if (res.status === 404) throw new SandboxNotFoundError(msg);
      throw new SandboxMatrixError(`API error (${res.status}): ${msg}`);
    }
    return data as T;
  }

  // Health & Version
  async health(): Promise<{ status: string }> {
    return this.request("GET", "/health");
  }

  async version(): Promise<VersionInfo> {
    return this.request("GET", "/version");
  }

  // Sandbox operations
  async createSandbox(name: string, blueprint: string, workspace?: string): Promise<Sandbox> {
    const body: Record<string, string> = { name, blueprint };
    if (workspace) body.workspace = workspace;
    const data = await this.request<any>("POST", "/sandboxes", body);
    return this.parseSandbox(data);
  }

  async getSandbox(name: string): Promise<Sandbox> {
    const data = await this.request<any>("GET", `/sandboxes/${encodeURIComponent(name)}`);
    return this.parseSandbox(data);
  }

  async listSandboxes(): Promise<Sandbox[]> {
    const data = await this.request<any[]>("GET", "/sandboxes");
    return data.map((s) => this.parseSandbox(s));
  }

  async startSandbox(name: string): Promise<void> {
    await this.request("POST", `/sandboxes/${encodeURIComponent(name)}/start`);
  }

  async stopSandbox(name: string): Promise<void> {
    await this.request("POST", `/sandboxes/${encodeURIComponent(name)}/stop`);
  }

  async destroySandbox(name: string): Promise<void> {
    await this.request("DELETE", `/sandboxes/${encodeURIComponent(name)}`);
  }

  async exec(name: string, command: string | string[]): Promise<ExecResult> {
    const cmd = typeof command === "string" ? ["sh", "-c", command] : command;
    return this.request("POST", `/sandboxes/${encodeURIComponent(name)}/exec`, { command: cmd });
  }

  async stats(name: string): Promise<Record<string, number>> {
    return this.request("GET", `/sandboxes/${encodeURIComponent(name)}/stats`);
  }

  // Snapshot operations
  async createSnapshot(name: string, tag?: string): Promise<{ snapshotId: string; tag: string }> {
    return this.request("POST", `/sandboxes/${encodeURIComponent(name)}/snapshots`, tag ? { tag } : {});
  }

  async listSnapshots(name: string): Promise<any[]> {
    return this.request("GET", `/sandboxes/${encodeURIComponent(name)}/snapshots`);
  }

  // Matrix operations
  async createMatrix(name: string, members: { name: string; blueprint: string }[]): Promise<any> {
    return this.request("POST", "/matrices", { name, members });
  }

  async getMatrix(name: string): Promise<any> {
    return this.request("GET", `/matrices/${encodeURIComponent(name)}`);
  }

  async listMatrices(): Promise<any[]> {
    return this.request("GET", "/matrices");
  }

  async startMatrix(name: string): Promise<void> {
    await this.request("POST", `/matrices/${encodeURIComponent(name)}/start`);
  }

  async stopMatrix(name: string): Promise<void> {
    await this.request("POST", `/matrices/${encodeURIComponent(name)}/stop`);
  }

  async destroyMatrix(name: string): Promise<void> {
    await this.request("DELETE", `/matrices/${encodeURIComponent(name)}`);
  }

  // Session operations
  async startSession(sandbox: string): Promise<any> {
    return this.request("POST", "/sessions", { sandbox });
  }

  async listSessions(sandbox?: string): Promise<any[]> {
    const path = sandbox ? `/sessions?sandbox=${encodeURIComponent(sandbox)}` : "/sessions";
    return this.request("GET", path);
  }

  async endSession(sessionId: string): Promise<void> {
    await this.request("POST", `/sessions/${encodeURIComponent(sessionId)}/end`);
  }

  private async requestRaw(method: string, path: string): Promise<string> {
    const url = `${this.baseURL}/api/v1${path}`;
    const headers: Record<string, string> = {};
    if (this.token) {
      headers["Authorization"] = `Bearer ${this.token}`;
    }

    const res = await fetch(url, { method, headers });
    if (!res.ok) {
      let msg: string;
      try {
        const data = await res.json();
        msg = data?.error ?? JSON.stringify(data);
      } catch {
        msg = await res.text();
      }
      if (res.status === 404) throw new SandboxNotFoundError(msg);
      throw new SandboxMatrixError(`API error (${res.status}): ${msg}`);
    }
    return res.text();
  }

  // File operations
  async writeFile(name: string, path: string, content: string): Promise<void> {
    const encoded = btoa(content);
    await this.request("PUT", `/sandboxes/${encodeURIComponent(name)}/files`, {
      path,
      content: encoded,
    });
  }

  async readFile(name: string, path: string): Promise<string> {
    return this.requestRaw(
      "GET",
      `/sandboxes/${encodeURIComponent(name)}/files?path=${encodeURIComponent(path)}`,
    );
  }

  async listFiles(name: string, path: string = "/"): Promise<FileInfo[]> {
    return this.request(
      "GET",
      `/sandboxes/${encodeURIComponent(name)}/files/list?path=${encodeURIComponent(path)}`,
    );
  }

  // Port operations
  async listPorts(name: string): Promise<PortMapping[]> {
    return this.request("GET", `/sandboxes/${encodeURIComponent(name)}/ports`);
  }

  // Code interpreter
  async interpret(
    name: string,
    language: string,
    code: string,
    timeout: number = 30,
  ): Promise<InterpretResult> {
    return this.request("POST", `/sandboxes/${encodeURIComponent(name)}/interpret`, {
      language,
      code,
      timeout,
    });
  }

  // Image operations
  async buildImage(blueprint: string): Promise<BuildResult> {
    return this.request("POST", "/images/build", { blueprint });
  }

  async listImages(): Promise<any[]> {
    return this.request("GET", "/images");
  }

  private parseSandbox(data: any): Sandbox {
    const metadata = data.metadata ?? {};
    const spec = data.spec ?? {};
    const status = data.status ?? {};
    return {
      name: metadata.name ?? "",
      state: status.state ?? "Unknown",
      blueprint: spec.blueprintRef ?? "",
      runtimeId: status.runtimeID ?? "",
      ip: status.ip,
      createdAt: metadata.createdAt,
    };
  }
}
